### PR TITLE
Networking enhancements Part 2: Dynamic backend

### PR DESCRIPTION
**Pull request by BradleyWire**
## What's new in this PR?

Part 2 of https://github.com/wireapp/wire-android/pull/2723

In this PR:

I have implemented a dynamic backend based on the preferences items provided to us by `Scala` when the app launches. This covers a dynamic `baseUrl` and certificate pins for now. 
#### APK
[Download build #1655](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1655/artifact/build/artifact/wire-dev-PR2724-1655.apk)
[Download build #1656](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1656/artifact/build/artifact/wire-dev-PR2724-1656.apk)
[Download build #1692](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1692/artifact/build/artifact/wire-dev-PR2724-1692.apk)
[Download build #1724](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1724/artifact/build/artifact/wire-dev-PR2724-1724.apk)
[Download build #1726](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1726/artifact/build/artifact/wire-dev-PR2724-1726.apk)

### CODE REVIEW

File: app/src/main/kotlin/com/waz/zclient/core/backend/di/BackendModule.kt [hunk 1]
```diff
@@ -1,10 +1,11 @@
 package com.waz.zclient.core.backend.di
 
-import com.waz.zclient.core.backend.BackendConfig
+import com.waz.zclient.core.backend.BackendClient
+import com.waz.zclient.storage.pref.backend.BackendPreferences
 import org.koin.core.module.Module
 import org.koin.dsl.module
 
 val backendModule: Module = module {
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/core/backend/di/BackendModule.kt]: I wonder if we switch to a custom backend for any of our customers how this config module will react?
Keep in mind that some of the customers don't use wire production backend to login. Try to investigate about that before merging this code.

Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/core/backend/di/BackendModule.kt]: Very good catch! This won't react at all unless they're storing the details in preferences. Will investigate this.

File: app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/BackendPrefsDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        )
+    )
+
+    fun getCustomBackendConfig(): Either<Failure, CustomBackendPrefResponse> =
+        if (customBackendConfig.isValid()) {
+            Either.Right(customBackendConfig)
+        } else Either.Left(InvalidBackendConfig)
+
+    fun updateCustomBackendConfig(configUrl: String, backendPrefResponse: CustomBackendPrefResponse) {
+        backendPreferences.customConfigUrl = configUrl
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/BackendPrefsDataSource.kt]: you can use `with(backendPreferences)` here

File: app/src/main/kotlin/com/waz/zclient/core/backend/usecase/CustomBackend.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.waz.zclient.core.backend.usecase
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/core/backend/usecase/CustomBackend.kt]: why it's below `/core/backend/usecase` ? Can't find any useCase here.

File: app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package com.waz.zclient.core.extension
+
+import com.waz.zclient.core.functional.Either
+
+/**
+ * Applies fnL if this is a Left or fnR if this is a Right.
+ * @see Left
+ * @see Right
+ */
+suspend fun <L, R, T> Either<L, R>.foldSuspendable(fnL: suspend (L) -> T?, fnR: suspend (R) -> T?): T? =
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt]: I think we can keep this inside the `Either` class itself as we are implementing this class. No need to add an Extension file for `Either`.

File: app/src/main/kotlin/com/waz/zclient/core/functional/FallbackOnFailure.kt [hunk 5]
```diff
@@ -12,7 +13,7 @@ import kotlinx.coroutines.runBlocking
  * If a [fallbackSuccessAction] is provided, it also invokes the action upon successful [fallbackAction].
  */
 data class FallbackOnFailure<R>(
-    private val primaryAction: suspend () -> Either<Failure, R>,
+    private val primaryAction: suspend () -> Either<Failure, Rx>,
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/core/functional/FallbackOnFailure.kt]: `Rx`?

File: app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt (hunk 4 above)
Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt]: You'll have to speak to @gizemb around this as we spoke about moving this out of the file.

Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt]: @gizemb why do we need an extension function for **Either** ? We usually implement extension function for classes that we don't own.

**Comment by BradleyWire** [comment]
build again

**Comment by BradleyWire** [comment]
build again

**Comment by BradleyWire** [comment]
build again

**Comment by BradleyWire** [comment]
build again

**Comment by BradleyWire** [comment]
build again

File: app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt (hunk 4 above)
Comment (mejdoo) [reply app/src/main/kotlin/com/waz/zclient/core/extension/Either.kt]: I moved this discussion to the next collective

File: app/src/main/kotlin/com/waz/zclient/core/backend/datasources/BackendDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    private fun getCustomBackendConfigRemotely(url: String) =
+        runBlocking {
+            remoteDataSource.getCustomBackendConfig(url).map {
+                backendMapper.toCustomBackend(it)
+            }
+        }
+
+    private fun getCustomBackendConfigLocally(): suspend () -> Either<Failure, CustomBackend> = {
+        runBlocking {
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/core/backend/datasources/BackendDataSource.kt]: I think you won't need runBlocking if you convert this method to suspend fun

File: app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/BackendPrefsDataSource.kt [hunk 7]
```diff
@@ -0,0 +1,39 @@
+package com.waz.zclient.core.backend.datasources.local
+
+import com.waz.zclient.core.exception.Failure
+import com.waz.zclient.core.exception.FeatureFailure
+import com.waz.zclient.core.functional.Either
+import com.waz.zclient.storage.pref.backend.BackendPreferences
+
+object InvalidBackendConfig : FeatureFailure()
+
+class BackendPrefsDataSource(private val backendPreferences: BackendPreferences) {
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/BackendPrefsDataSource.kt]: can be `BackendLocalDataSource`

File: app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/CustomBackendPrefResponse.kt [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.waz.zclient.core.backend.datasources.local
+
+data class CustomBackendPrefResponse(
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/core/backend/datasources/local/CustomBackendPrefResponse.kt]: we don't use "response" for database models, so I think there's no need to use here.

**Comment by mejdoo** [comment]
The app is crashing again when I press **About** button in settings screen. Log attached...
[di_crash.pdf](https://github.com/wireapp/wire-android/files/4402947/di_crash.pdf)

**Comment by BradleyWire** [comment]
> The app is crashing again when I press **About** button in settings screen. Log attached...
> [di_crash.pdf](https://github.com/wireapp/wire-android/files/4402947/di_crash.pdf)

Can you tell me

File: app/src/main/kotlin/com/waz/zclient/core/network/backend/CustomBackendInterceptor.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+class CustomBackendInterceptor(
+    private val backendRepository: BackendRepository,
+    private val configUrl: String
+) : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response =
+        if (configUrl.isEmpty()) {
+            chain.proceed(chain.request())
+        } else {
+            val request = buildConfigRequest(chain)
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/core/network/backend/CustomBackendInterceptor.kt]: it's clever but is it ok performance-wise? seems like it's accessing to sharedPreferences every time it makes a request, which may delay the network calls for some time.

Comment (BradleyWire) [reply app/src/main/kotlin/com/waz/zclient/core/network/backend/CustomBackendInterceptor.kt]: I'll investigate this tomorrow.

**Pull request merged**